packages/@aws-cdk/aws-iam/lib/optimize-statements.ts

```diff
@@ -0,0 +1,286 @@
+   */
+  supercedes(other: this): boolean;
+}
+
+/**
+ * Entities that can be merged.
+ */
+interface Mergeable {
+  /**
+   * Checks whether an entity is compatible for merging in this one.
+   * @param other the other entity to check.
+   * @returns ``true`` if ``other`` can safely be merged into this entity.
+   */
+  mergeable(other: this): boolean;
+  /**
+   * Merges ``other`` in this entity. It is the caller's responsibility to have
+   * verified the safety of doing so by having called ``mergeable`` before.
+   * @param other the entity to merge in.
+   */
+  merge(other: this): void;
+}
+function _isMergeable<T>(x: T): x is T & Mergeable {
+  return x != null
+    && typeof (x as any).mergeable === 'function'
+    && typeof (x as any).merge === 'function';
+}
+
+/**
+ * Removes duplicate entities from a list.
+ * @param array the list to be de-duplciated.
+ * @returns a new list containing the minimum set of entities.
+ */
+function _deduplicate<T extends Deduplicable>(array: T[]): T[] {
+  const result = new Array<T>();
+  for (let i = 0 ; i < array.length ; i++) {
+    const element = array[i];
+    let redundant = false;
+    for (let j = i + 1 ; j < array.length ; j++) {
+      const candidate = array[j];
+      if (candidate.supercedes(element)) {
+        redundant = true;
+        break;
+      } else if (_isMergeable(candidate) && _isMergeable(element) && candidate.mergeable(element)) {
+        candidate.merge(element);
+        redundant = true;
+        break;
+      }
+    }
+    if (!redundant) { result.push(element); }
+  }
+  return result;
+}
+
+/**
+ * Checks if two entities are equal.
+ * @param e1 one of the entities.
+ * @param e2 the other entity.
+ * @returns true if both entities are identical.
+ */
+function _deepEqual(e1: unknown, e2: unknown): boolean {
+  if (typeof e1 !== typeof e2) { return false; }
+  if (Array.isArray(e1)) {
+    if (!Array.isArray(e2)) { return false; }
+    if (e1.length !== e2.length) { return false; }
+    return e1.find(e => e2.find(c => _deepEqual(e, c)) != null) == null;
+  }
+  if (typeof e1 === 'object' && typeof e2 === 'object') {
+    if (e1 == null || e2 == null) { return e1 === e2; }
+    const e1k = Object.keys(e1 as any);
+    const e2k = new Set(Object.keys(e2 as any));
+    if (e1k.length !== e2k.size) { return false; }
+    for (const key of e1k) {
+      if (!e2k.has(key)) { return false; }
+      if (!_deepEqual((e1 as any)[key], (e2 as any)[key])) { return false; }
+    }
+    return true;
+  }
+  return e1 === e2;
+}
+
+/**
+ * Makes a regular expression from a glob pattern.
+ *
+ * @param glob a glob pattern that uses '*' to denote wild cards.
+ *
+ * @returns ``undefined`` if ``glob`` is not a string, otherwise a ``RegExp`` corresponding to the glob pattern.
+ */
+function _toRegExp(glob: unknown): RegExp | undefined {
+  if (typeof glob !== 'string') { return undefined; }
+  const parts = glob.split('*');
+  return new RegExp(`^${parts.map(_quote).join('.*')}\$`);
```

Comment:
`.*` ? Don't know that that is true. For example, ARN globs don't extend over `:`. So at the very least, I think it should be:

```
[^:]*
```

Reply:
Also, not sure that the string substitution format helps here.

```
'^' + parts.map(_quote).join('.*') + '$'
```

Would have been just as readable, if not more so.

Reply:
Owhhh yeah good catch!